File: componenthub_mcp/coordinator.py
"""Search coordinator: fan searches out to providers, collect, normalize, merge.

No ranking — results are merged by MPN and returned as-is. The connected LLM
decides how to compare or order them. Everything is in-memory per request.
"""

import asyncio
import time
from collections import deque
from typing import Any

from .models import Capability, ComponentResult, SearchQuery
from .providers import ProviderError, all_providers, get_provider

# Tiny search history (spec: store search, timestamp, providers — never components).
_history: deque[dict[str, Any]] = deque(maxlen=50)


def resolve_providers(names: list[str] | None, capability: Capability) -> tuple[list, list[dict]]:
    """Pick providers that can answer this capability; report ones that can't."""
    skipped: list[dict] = []
    if names:
        candidates = [get_provider(n) for n in names]
    else:
        candidates = list(all_providers().values())
    usable = []
    for p in candidates:
        if capability not in p.capabilities:
            if names:  # only report if explicitly requested
                skipped.append({"provider": p.name, "reason": f"does not support {capability.value}"})
            continue
        if not p.is_configured():
            skipped.append({"provider": p.name, "reason": p.missing_config() or "not configured"})
            continue
        usable.append(p)
    return usable, skipped


def _merge(results: list[ComponentResult]) -> list[ComponentResult]:
    """Merge results from different providers that refer to the same MPN."""
    merged: dict[str, ComponentResult] = {}
    for r in results:
        key = r.mpn.strip().upper()
        if not key:
            merged[id(r).__str__()] = r
            continue
        if key in merged:
            existing = merged[key]
            existing.offers.extend(r.offers)
            existing.manufacturer = existing.manufacturer or r.manufacturer
            existing.description = existing.description or r.description
            existing.category = existing.category or r.category
            existing.package = existing.package or r.package
            existing.datasheet_url = existing.datasheet_url or r.datasheet_url
            existing.image_url = existing.image_url or r.image_url
        else:
            merged[key] = r.model_copy(deep=True)
    return list(merged.values())


async def search(query: SearchQuery, provider_names: list[str] | None) -> dict[str, Any]:
    providers, skipped = resolve_providers(provider_names, Capability.SEARCH)
    errors: list[dict] = list(skipped)

    async def run_one(p) -> list[ComponentResult]:
        try:
            return await p.search(query)
        except ProviderError as e:
            errors.append({"provider": p.name, "reason": str(e)})
        except Exception as e:  # network failures etc. — never fatal
            errors.append({"provider": p.name, "reason": f"unexpected error: {e}"})
        return []

    batches = await asyncio.gather(*(run_one(p) for p in providers))
    flat = [r for batch in batches for r in batch]
    merged = _merge(flat)

    _history.appendleft(
        {
            "query": query.keyword,
            "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            "providers": [p.name for p in providers],
            "result_count": len(merged),
        }
    )

    return {
        "results": [r.model_dump(exclude_none=True) for r in merged],
        "providers_searched": [p.name for p in providers],
        "providers_skipped": errors,
        "note": "Results are unranked and unfiltered — compare and rank them yourself as needed.",
    }


def history() -> list[dict[str, Any]]:
    return list(_history)

File: src/lib/domain/coordinator.ts
/**
 * Search coordinator — TS port of coordinator.py.
 * Fan searches out to providers, collect, normalize, merge by MPN. No ranking.
 *
 * NOTE: `history` is in-memory and therefore ephemeral on serverless (one warm
 * instance only). The web app persists real per-user history in Supabase via
 * lib/data/history.ts. This ring buffer exists for MCP `recent_searches` parity.
 */
import {
  Capability,
  type ComponentResult,
  excludeNone,
  type SearchQuery,
} from "./models";
import { allProviders, getProvider, Provider, ProviderError } from "./providers/registry";

interface SkipReason {
  provider: string;
  reason: string;
}
interface HistoryEntry {
  query: string;
  timestamp: string;
  providers: string[];
  result_count: number;
}

const HISTORY_MAX = 50;
const historyRing: HistoryEntry[] = [];

export function resolveProviders(
  names: string[] | null | undefined,
  capability: Capability,
): { usable: Provider[]; skipped: SkipReason[] } {
  const skipped: SkipReason[] = [];
  const candidates = names && names.length
    ? names.map((n) => getProvider(n))
    : [...allProviders().values()];
  const usable: Provider[] = [];
  for (const p of candidates) {
    if (!p.capabilities.has(capability)) {
      if (names && names.length) {
        skipped.push({ provider: p.name, reason: `does not support ${capability}` });
      }
      continue;
    }
    if (!p.isConfigured()) {
      skipped.push({ provider: p.name, reason: p.missingConfig() ?? "not configured" });
      continue;
    }
    usable.push(p);
  }
  return { usable, skipped };
}

/** Merge results from different providers that refer to the same MPN. */
function merge(results: ComponentResult[]): ComponentResult[] {
  const merged = new Map<string, ComponentResult>();
  let anon = 0;
  for (const r of results) {
    const key = r.mpn.trim().toUpperCase();
    if (!key) {
      merged.set(`__anon_${anon++}`, r);
      continue;
    }
    const existing = merged.get(key);
    if (existing) {
      existing.offers.push(...r.offers);
      existing.manufacturer = existing.manufacturer || r.manufacturer;
      existing.description = existing.description || r.description;
      existing.category = existing.category || r.category;
      existing.package = existing.package || r.package;
      existing.datasheet_url = existing.datasheet_url || r.datasheet_url;
      existing.image_url = existing.image_url || r.image_url;
      if (r.specifications) {
        existing.specifications ??= {};
        for (const [k, v] of Object.entries(r.specifications)) {
          if (!(k in existing.specifications)) existing.specifications[k] = v;
        }
      }
    } else {
      merged.set(key, structuredClone(r));
    }
  }
  return [...merged.values()];
}

export interface SearchResponse {
  results: ComponentResult[];
  providers_searched: string[];
  providers_skipped: SkipReason[];
  note: string;
}

export async function search(
  query: SearchQuery,
  providerNames: string[] | null | undefined,
): Promise<SearchResponse> {
  const { usable, skipped } = resolveProviders(providerNames, Capability.SEARCH);
  const errors: SkipReason[] = [...skipped];

  const runOne = async (p: Provider): Promise<ComponentResult[]> => {
    try {
      return await p.search(query);
    } catch (e) {
      const reason =
        e instanceof ProviderError ? e.message : `unexpected error: ${String(e)}`;
      errors.push({ provider: p.name, reason });
      return [];
    }
  };

  const batches = await Promise.all(usable.map(runOne));
  const flat = batches.flat();
  const merged = merge(flat);

  historyRing.unshift({
    query: query.keyword,
    timestamp: new Date().toISOString().replace(/\.\d+Z$/, "Z"),
    providers: usable.map((p) => p.name),
    result_count: merged.length,
  });
  if (historyRing.length > HISTORY_MAX) historyRing.length = HISTORY_MAX;

  return {
    results: merged.map((r) => excludeNone(r)),
    providers_searched: usable.map((p) => p.name),
    providers_skipped: errors,
    note: "Results are unranked and unfiltered — compare and rank them yourself as needed.",
  };
}

export function history(): HistoryEntry[] {
  return [...historyRing];
}
